component/module/git/internal/vcs/git.go

```diff
@@ -58,15 +58,15 @@ func NewGitRepo(ctx context.Context, storagePath string, opts GitRepoOptions) (*
 	}
 
 	// Fetch the latest contents. This may be a no-op if we just did a clone.
-	err = repo.FetchContext(ctx, &git.FetchOptions{
+	fetchRepoErr := repo.FetchContext(ctx, &git.FetchOptions{
 		RemoteName: "origin",
 		Force:      true,
 		Auth:       opts.Auth.Convert(),
 	})
-	if err != nil && !errors.Is(err, git.NoErrAlreadyUpToDate) {
-		return nil, UpdateFailedError{
+	if fetchRepoErr != nil && !errors.Is(fetchRepoErr, git.NoErrAlreadyUpToDate) {
+		err = UpdateFailedError{
 			Repository: opts.Repository,
-			Inner:      err,
+			Inner:      fetchRepoErr,
```

Comment:
Hm, I'd rather the error handler here return instead of falling back to Reset, that way we don't touch the repo at all in case a pull fails; something like

```go
if err != nil && !errors.Is(err, git.NoErrAlreadyUpToDate) {
  workTree, workTreeErr := repo.Worktree()
  if workTreeErr != nil {
    return nil, workTreeErr  
  }

  return &GitRepo{
    opts: opts, 
    repo: repo, 
    workTree: workTree, 
  }, UpdateFailedError{
    Repository: opts.Repository,
    Inner: err, 
  }
}
```

(This can probably be refactored a bit so the GitRepo object is constructed before doing a pull to reduce duplication with the successful path)

Reply:
Good point here. I've amended in latest commit.